### DISCUSSION TITLE
Combining prompt=none with PKCE authentication code flow … possible?

### DISCUSSION THREAD

**Discussion by jpommerening** [keycloak/keycloak#28667] [Q&A] [answered]
Similar question to #9246:

Is it possible to have an OpenID Connect login using the authentication code flow with `prompt=none`?

My scenario differs slightly from #9246 in that I'm not using the `id_token_hint`, but am trying to rely on an already established session stored in a cookie.

I'm using (in Keycloak terms) a non-confidential client. This client is supposed to authenticate using the PKCE challenge algorithm, using `oidc-client-ts` as a client library.

It is the scenario described [here](https://github.com/authts/oidc-client-ts/blob/main/docs/protocols/silent-refresh-token-in-iframe-flow.md) (incl. helpful flowchart) that I'm trying to accomplish.

The client that is configured in Keycloak has only the "Standard Flow" enabled and requires no consent.

Here is what happens:

- When I try the "silent" login with `prompt=none` in a hidden `iframe`, Keycloak redirects with a [`error=login_required` response](https://openid.net/specs/openid-connect-core-1_0.html#AuthError).

- When I try the same setup using a redirect in the main window and `prompt=login` (i.e., [this](https://github.com/authts/oidc-client-ts/blob/main/docs/protocols/authorization-code-grant-with-pkce.md)), I get redirected straight back to my application with a successful login (and no login form). I take that as an indication that I already have a working session.

**Comment by justin-tay** [comment, marked as answer]
With browsers dropping support for third-party cookies it is likely that the session cookie isn't being sent in when in the `iframe`.

You can see https://www.keycloak.org/docs/latest/securing_apps/#_modern_browsers for some information.

**Reply by jonkoops** [reply]
Yeah, this could be part of it. In recent versions of Keycloak we've updated the code in the iframe that handles the Silent Auth to [emit an `error`](https://github.com/keycloak/keycloak/blob/92bcd2645c35575c756a01fe39462f40633c65ee/services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html#L32-L73) state when we cannot access the cookies.

**Reply by jpommerening** [reply]
Oh wow that helped me a lot!
I didn't think of that _at all_!

Trying it with Safari (& making sure I can capture the redirect) I can see that it doesn't send any cookies …
Of course its telling me `login_required` …
<img width="791" alt="image" src="https://github.com/keycloak/keycloak/assets/748636/be76f4e8-9955-4029-9d48-8cb0f64f5827">

Since the refresh with `id_token_hint` works, and it's only the initial visit that needs the login, my fallback strategy will rely on the "foreground" redirect and I'll just try to start that as soon as possible to make it less jarring :)

Do you have any other recommendations?

**Reply by jonkoops** [reply]
> my fallback strategy will rely on the "foreground" redirect and I'll just try to start that as soon as possible to make it less jarring :)

That's pretty much what we do for Keycloak JS as well. It's not the best UX, but until we work something out with FedCM (#16834) there isn't much we can do to make it prettier.
